client/asset/dcr/dcr.go

```diff
@@ -820,35 +818,11 @@ func (dcr *ExchangeWallet) split(value uint64, lots uint64, coins asset.Coins, i
 		dcr.log.Errorf("error locking funding coin from split transaction %s", op)
 	}
 
-	// NOTE: Can't return coins yet, because dcrwallet doesn't recognize them as
-	// spent immediately, so subsequent calls to FundOrder might result in a
-	// `-4: rejected transaction: transaction in the pool already spends the
-	// same coins` error.
-	// TODO: Could this cause balance report inaccuracy, where the locked atoms
-	// returned by dcrwallet includes both the funding coins and the split tx
-	// output?
-	// E.g. total of 100 DCR (locked) split to produce a desired output of 40 DCR
-	// that is also locked:
-	// dcr balance before split, total = 200 DCR, locked = 0
-	// dcr balance after split, total = 199 DCR (-split tx fee), locked = 140 DCR
-	// If this inaccurate report is a possibility, might be better to unlock the
-	// funding coins sooner rather than later and prevent FundOrder double spends
-	// by checking listunspent results against `dcr.splitFunds`.
-	// If the split fund txout is later unlocked without being spent, the initial
-	// funding coins will be deleted from dcr.splitFunds, making them re-spendable
-	// by dcr.FundOrder.
-	// // Unlock the spent coins.
-	// err = dcr.returnCoins(coins)
-	// if err != nil {
-	// 	dcr.log.Errorf("error locking coins spent in split transaction %v", coins)
-	// }
-
-	// Associate the funding coins with the split tx output, so that the coins
-	// can be unlocked when the swap is sent. The helps to deal with a timing
-	// issue with dcrwallet where listunspent might still return outputs that
-	// were just spent in a transaction broadcast with sendrawtransaction.
-	// Instead, we'll keep them locked until the split output is spent.
-	dcr.splitFunds[op.pt] = fundingCoins
+	// Unlock the spent coins.
+	err = dcr.returnCoins(coins)
+	if err != nil {
+		dcr.log.Errorf("error returning coins spent in split transaction %v", coins)
+	}
```

Comment:
This reminds me that we have to explicitly unlock coins that we locked after they're spent otherwise listlockunspent might continue to return them.

dcr.sendRegFee and dcr.Withdraw both call dcr.fund to select input coins, dcr.fund locks the selected coins before returning them but dcr.sendRegFee and dcr.Withdraw do not unlock the coins after spending the coins.

First thought was make `dcr.fund` accept a bool parameter that determines whether to lock the selected coins - basically, don't lock if we're gonna be spending them right away but inbetween the brief moment between when the selected coins are returned and when they're spent, listunspent would return them, because the fundingMtx lock is released when dcr.fund returns. Also, another time when we should not lock the selected coins are when we plan to spend them in a split tx right away but we may end up not spending them (if it's uneconomical to use a split) and then we'd need to lock the coins.

The plenty rant above is to lay the groundwork for the changes coming in my next commit: don't lock coins in dcr.fund or even dcr.split. Let callers lock the fundingMtx before calling dcr.fund and dcr.split and only release the lock after either spending the selected coins or locking them.

Reply:
Okay, the only issue is here
> dcr.sendRegFee and dcr.Withdraw both call dcr.fund to select input coins, dcr.fund locks the selected coins before returning them but dcr.sendRegFee and dcr.Withdraw do not unlock the coins after spending the coins.

which does not seem to have been an issue before now.

Skipping this change for this PR. It's a better candidate for another PR I have in the works.